src/core/TokenGraphemeSplitter.ts

```diff
@@ -0,0 +1,165 @@
+import { escapeRegExp } from "lodash";
+import { Disposable } from "../ide/ide.types";
+import { Graph, TokenHatSplittingMode } from "../typings/Types";
+import { Notifier } from "../util/Notifier";
+import { matchAll } from "../util/regex";
+
+const KNOWN_SYMBOLS = [
+  "!",
+  "#",
+  "$",
+  "%",
+  "&",
+  "'",
+  "(",
+  ")",
+  "*",
+  "+",
+  ",",
+  "-",
+  ".",
+  "/",
+  ":",
+  ";",
+  "<",
+  "=",
+  ">",
+  "?",
+  "@",
+  "[",
+  "\\",
+  "]",
+  "^",
+  "_",
+  "`",
+  "{",
+  "|",
+  "}",
+  "~",
+  "£",
+  '"',
+];
+const KNOWN_SYMBOL_REGEXP = KNOWN_SYMBOLS.map(escapeRegExp).join("|");
+
+const KNOWN_GRAPHEME_REGEXP = [
+  "\\p{L}\\p{M}*",
+  "[0-9]",
+  KNOWN_SYMBOL_REGEXP,
+].join("|");
+const KNOWN_GRAPHEME_MATCHER = new RegExp(`^${KNOWN_GRAPHEME_REGEXP}$`, "u");
+
+export const UNKNOWN = "[unk]";
+
+const SPLIT_REGEX = /\p{L}\p{M}*|\P{L}/gu;
+
+export class TokenGraphemeSplitter {
+  private disposables: Disposable[] = [];
+  private algorithmChangeNotifier = new Notifier();
+  private tokenHatSplittingMode!: TokenHatSplittingMode;
+
+  constructor(private graph: Graph) {
+    graph.ide.disposeOnExit(this);
+
+    this.updateTokenHatSplittingMode =
+      this.updateTokenHatSplittingMode.bind(this);
+    this.getTokenGraphemes = this.getTokenGraphemes.bind(this);
+
+    this.updateTokenHatSplittingMode();
+
+    this.disposables.push(
+      // Notify listeners in case the user changed their token hat splitting
+      // setting.
+      this.graph.ide.configuration.onDidChangeConfiguration(
+        this.updateTokenHatSplittingMode
+      )
+    );
+  }
+
+  private updateTokenHatSplittingMode() {
+    const { accentsToPreserve, symbolsToPreserve, ...rest } =
+      this.graph.ide.configuration.getOwnConfiguration<TokenHatSplittingMode>(
+        "tokenHatSplittingMode"
+      )!;
+
+    this.tokenHatSplittingMode = {
+      accentsToPreserve: accentsToPreserve.map((grapheme) =>
+        grapheme.toLowerCase().normalize("NFC")
```

Comment:
This is a bit awkward.  We still letters to lowercase here even if they have set `preserveCase=true`.  I wasn't entirely sure what to do in that case.  To me it seemed that even if they want to preserve case, it's unlikely they'd want to preserve accents for lowercase version of letter but not uppercase

Fwiw we can get around this issue entirely by only paying attention to the accents themselves in this string.  Basically do NFD normalisation and then keep only the diacritics (kind of the opposite of what we do to strip the marks below).  That would mean that they can't keep a particular diacritic for one char but not another, ie preserve accent for `ö` but not `ä`

cc/ @AndreasArvidsson @auscompgeek

Reply:
In principle I agree too. I wonder if there are any characters that have accents where `upper(lower(char)) != char`.

Reply:
@auscompgeek what would be the problem if such characters existed? I don't believe I'm relying on that fact anywhere, but definitely might be missing something

Reply:
I'm probably overthinking it, but just thinking of characters that behave like "ß" for example (but in the other direction for this particular scenario). In such a case it'd probably be a mistake to treat the lowercase and uppercase of such characters as the same.

Reply:
huh interesting.  I guess for these cases one could just use the other setting (symbols to preserve)? An example use case for that was `Σ` vs `σ`, where you'd probably want different spoken forms for those if you're just using them for math, even if you prefer case normalisation for Latin alphabet

Reply:
Ah yeah, makes sense. :+1: